node/blockchain/models/block/manager.py

```diff
@@ -23,14 +23,15 @@ def add_block_from_signed_change_request(
         signing_key: Optional[SigningKey] = None,
         validate=True
     ) -> 'Block':
-        if validate:
-            signed_change_request.validate_business_logic(blockchain_facade)
+        with TableLocker(self.model):
+            if validate:
+                signed_change_request.validate_business_logic(blockchain_facade)
 
-        block_message = BlockMessage.create_from_signed_change_request(signed_change_request, blockchain_facade)
-        # no need to validate the block message since we produced a valid one
-        return self.add_block_from_block_message(
-            block_message, blockchain_facade, signing_key=signing_key, validate=False
-        )
+            block_message = BlockMessage.create_from_signed_change_request(signed_change_request, blockchain_facade)
+            # no need to validate the block message since we produced a valid one
+            return self.add_block_from_block_message(
+                block_message, blockchain_facade, signing_key=signing_key, validate=False
+            )
 
     def add_block_from_block_message(
```

Comment:
the lock must be here and on add_block() method as well.

it should be possible to call `add_block_from_block_message(other arguments, expect_locked=True)`, so the decorator swallows the `expect_locked` argument and just checks that the lock is already set

See `Have a single decorator with an argument that it intercepts and changes behavior at run-time: expect_locked` from the task

Reply:
I added "expect_locked" argument to "add_block_from_block_message". Also I decorated "BlockManager.add_block" method by decorator "ensure_locked".